include/ruby/internal/arithmetic/long.h

```diff
@@ -114,11 +114,11 @@ RB_INT2FIX(long i)
 
     /* :NOTE: VALUE can be wider than long.  As j being unsigned, 2j+1 is fully
      * defined. Also it can be compiled into a single LEA instruction. */
-    const unsigned long j = i;
+    const unsigned long j = (unsigned long)i;
```

Comment:
For C++ compilation, we need to wrap casts with `RBIMPL_CAST` in the public headers.

```suggestion
    const unsigned long j = RBIMPL_CAST((unsigned long)i);
```

Same for all the rest.